Diamond fixes

Wiebke opened the pull request:
This PR encompasses a number of fixes and extensions we came across while testing:

- Exports mask indices as a list of integers rather than a list of strings (this is also the way the model saves mask indices to Tiled)
- Adds the `num_classes` parameter to the inference job
- Extends the colormap by a gray color for unlabeled data and fixes a small bug in the color map scale when classes where deleted
- Autoscales the data with percentiles
- Passes the environment variables `RESULTS_DIR` as a new io parameter to the model as `models_dir`

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207690415598274

Wiebke commented:
Two additional changes that I just added here: 
- Fix a bug with the result overlay toggle not correctly updating when results come in
- Disable keybinds for annotation and switching slices when the model controls accordion is opened. This means typing letters or numbers in those fields no longer leads to the annotation mode changing or slices updating unintentionally.

taxe10 commented on callbacks/control_bar.py:
```diff
@@ -183,6 +192,8 @@ def annotation_mode(
     if generate_modal_opened or any(edit_modal_opened):
         # user is going to type on this page (on a modal) and we don't want to trigger this callback using keys
         raise PreventUpdate
+    if control_accordion_state is not None and "run-model" in control_accordion_state:
```
Disabling key binds while interacting with the model configuration tab makes a lot of sense. While testing this at my end, I accidentally left the model configuration tab open while meaning to do some more annotations. Here I noticed that the key binds were off because the model configuration tab was still open.
My suggestion is to maybe allow only 1 tab/accordion item in the left panel to be active at the time. So when the user selects one, the other close automatically (between model and annotations tabs).

Wiebke replied:
I think I would prefer classes to be visible while interacting with the result overlay. Perhaps we want to explore splitting out job submission and/or results into an additional tab?

taxe10 commented on utils/plot_utils.py:
```diff
@@ -182,17 +182,23 @@ def generate_segmentation_colormap(all_annotations_data):
+    # (annotation_class["class_id"] for annotation_class in all_annotations_data)
+    # may have larger ids than num_classes -1 and gaps
+    # this is not the case for segmented results or masks
+    max_class_id = len(all_annotations_data) - 1
```
Not sure if this is related to this change, but the segmented mask have missing classes when loading previously segmented results if I don't manually pre-setup the annotations classes to match the classes in the mask.
For example, if I have previously segmented a dataset with 2 classes. After reloading, there is only 1 default class. If I select a trained model and proceed to "view segmentation overlay", the segmented mask shows 1 class only instead of 2.

Wiebke replied:
This is unrelated to the change, but something we should indeed address.

Currently it is assumed that the number of classes listed in the interface does match with the classes of the segmentation. While this does enable the user to e.g. change colors of the overlay *and* annotations through simply interacting with the existing interface for classes, it can cause inconsistencies if the user e.g. deletes a class, or reloads the interface, annotates with less classes than before, but then still loads an older result.

For previously segmented results, we could choose to load class names, class colors, and annotations (along with model parameters) from the segmented result instead.

taxe10 replied:
Agreed. Moving this to an issue.

taxe10 commented on utils/plot_utils.py:
```diff
@@ -182,17 +182,23 @@ def generate_segmentation_colormap(all_annotations_data):
         annotation_class["color"] for annotation_class in all_annotations_data
     ]
+    # Add color for unlabeled pixels
+    color_list = ["#D3D3D3"] + color_list
```
Is this function used for segmentation results only? If so, is it possible to have unlabeled pixels?

Wiebke replied:
As far as I understand, this is possible in the case of ensemble methods, and perhaps even with certain configurations of `qlty` when no label can be conclusively assigned (due to conflicting or low probability assignment).

taxe10 commented on callbacks/segmentation.py:
```diff
@@ -427,37 +438,53 @@ def populate_segmentation_results(
+        train_job_id, project_name, "training"
+    )
+    if segment_job_id is not None:
+        results_link = os.path.join(RESULTS_DIR, segment_job_id, "results.html")
```
The current version in [mlex_dlsia](https://github.com/mlexchange/mlex_dlsia_segmentation_prototype) has setup the path to the DVC report to `dvc_metrics/report.html`
I made that change at my end to test it, but I am unable to see the DVC report from the browser. The URL updates to: `http://localhost:8075/path/to/report.html`, with `path/to/report.htm` being the correct path to the report. I wonder if there is a routing issue to the html file.

Wiebke replied:
It appears that linking to a local file is not permitted in most browsers due to security concerns and requires to start the browser in a certain way, or use of an extension. We may have to explore other means to show these metrics. 

Copying the same url in the tab, pasting it, then pressing enter should make the report show up though.

taxe10 replied:
That makes sense. Just a minor comment here is that the path to the dvc report has been updated in the latest version of dlsia from `results.html` to `dvc_metrics/report.html`, as detailed in the following [link](https://github.com/mlexchange/mlex_dlsia_segmentation_prototype/blob/e76c22afe25facefd490e40ce67647f85aa29474/src/train.py#L118).
```suggestion
        results_link = os.path.join(RESULTS_DIR, segment_job_id, "dvc_metrics/report.html")
```

taxe10 reviewed (changes requested):
Changes looks good. I left some minor comments in the PR.

While testing the app locally, I noticed the following as well, but I am not sure if they are related to this PR:

1. The `1-9` key binds are not working at my end
2. After making an annotation, I cannot switch to other drawing patterns (e.g. rectangular, circular, etc)
3. I found the following traceback error while using `pan and zoom`
```
Traceback (most recent call last):
  File "/Users/tanny/miniconda3/envs/segmentation/lib/python3.11/site-packages/flask/app.py", line 2529, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tanny/miniconda3/envs/segmentation/lib/python3.11/site-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tanny/miniconda3/envs/segmentation/lib/python3.11/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tanny/miniconda3/envs/segmentation/lib/python3.11/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tanny/miniconda3/envs/segmentation/lib/python3.11/site-packages/dash/dash.py", line 1297, in dispatch
    ctx.run(
  File "/Users/tanny/miniconda3/envs/segmentation/lib/python3.11/site-packages/dash/_callback.py", line 437, in add_context
    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/tanny/Documents/Coding/Direct/mlex_highres_segmentation/callbacks/control_bar.py", line 254, in annotation_mode
    shape["editable"] = trigger != "pan-and-zoom" and pan_and_zoom > 0
                                                      ^^^^^^^^^^^^^^^^
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

A very minor suggestion: `CrossEntropyLoss` should probably be the default loss selection. I accidentally ran the ML job with default parameters and there was an error due to MSELoss not having a `weight` parameter. I know this is related to our previous discussion about having some "parameter dependency". While we come up with a solution for this, I wonder if we should just remove the losses that do not take the `weight` parameter.

Wiebke commented:
Great catches! The `1-9` key binds were indeed not working due to a mismatch in ordering of elements in the callback decorator and arguments in the callback function. I also fixed the issue with pan and zoom, made `CrossEntropyLoss` the default loss, and fixed a minor bug that prevented the mask export from functioning.

I could not reproduce the issue regarding switching annotation mode, but would suspect it was related to the tow other issues. Could you retest?

taxe10 reviewed (commented):
Changes work great at my end. I made some very minor comments.

Wiebke commented on assets/models.json:
```diff
@@ -137,7 +137,7 @@
                     "title": "Criterion",
                     "param_key": "criterion",
-                    "value": "MSELoss",
+                    "value": "CrossEntropyLoss",
```
Done!
